### PR TITLE
add async support

### PR DESCRIPTION
**Pull request by katosad**
Hello @omarish! My org has been using your package to organize our business logic in our Flask apps and are big fans. Now, we're building a new app using FastAPI, which will be async. We'd like to continue using mutations to organize our functionality but we need a version that supports async `execute()` methods. I took a stab at adding an AsyncMutations class (and corresponding tests), but please let me know if you have any concerns/suggestions. Thanks so much!

Screen Shot 2023-02-10 at 4 48 17 PM

### CODE REVIEW

File: mutations/core.py [hunk 1]
```diff
@@ -131,3 +131,21 @@ def validate(cls, raise_on_error=False, **kwargs):
...
+        instance = cls(cls.__name__, inputs = kwargs)
+        is_valid, error_dict = instance._validate()
+
+        if not is_valid:
+            if raise_on_error:
+                raise error.MutationFailedValidationError(error_dict)
+            else:
+                return Result(success=False, return_value=None, errors=error_dict)
+
+        result = await instance.execute()
```
Comment (katosad) [inline mutations/core.py]: the only difference between `Mutation` and `AsyncMutation` is that we await the `execute()` function

File: requirements.txt [hunk 2]
```diff
@@ -1,7 +1,7 @@
 appnope==0.1.0
 astroid==1.6.1
-attrs==17.4.0
-certifi==2018.4.16
+attrs==22.2.0
+certifi==2022.12.7
```
Comment (katosad) [inline requirements.txt:4]: some of these updates pull in the dependabot changes - others are needed for compatibility with `pytest_asyncio`

**Comment by katosad** [comment]
@omarish @jourdanrodrigues FYI

**Comment by omarish** [comment]
Hi there, thanks for the PR and apologies for the delayed response. I'll have a close look in the next day or two.

**Comment by katosad** [comment]
> 

no worries at all and thanks so much :)

File: mutations/core.py (hunk 1 above)
Comment (omarish) [reply mutations/core.py]: Nice, this looks good. One question I've been thinking about and curious to hear your take.. what if we were to do it like this (pseudocode):

```python
import asyncio
import inspect

...

class Mutation():
    def run(cls, raise_on_error=False, loop=None, **kwargs):
        if inspect.isawaitable(...):
            result = asyncio.run(...)
        else:
            ...
```

I think the way you're proposing is more obvious and probably a better developer experience. What do you think?

**Comment by omarish** [comment]
Looks good! I left one question inline with the code. Let me know what you think.

File: mutations/core.py (hunk 1 above)
Comment (katosad) [reply mutations/core.py]: Oh I like that!! it'd be nice to have just one class that automatically handles sync vs async invocations. Will update my PR :) thanks for the suggestion!

File: mutations/core.py [hunk 3]
```diff
@@ -102,6 +116,17 @@ def _get_input(self, field):
...
+        try:
+            loop = asyncio.get_running_loop()
+        except RuntimeError:
+            loop = None
+
+        if loop and loop.is_running():
+            return RunThread(self.execute).result
+        else:
+            return asyncio.run(self.execute())
+
```
Comment (katosad) [inline mutations/core.py:129]: need to check if any event loops are already running  - https://stackoverflow.com/questions/55409641/asyncio-run-cannot-be-called-from-a-running-event-loop-when-using-jupyter-no

File: mutations/core.py (hunk 1 above)
Comment (katosad) [reply mutations/core.py]: @omarish updated! tests pass as before

**Comment by katosad** [comment]
hi @omarish just wanted to bump this once more - lmk if you have any questions/concerns! Thanks :)

**Pull request closed without merging**